Email with 50MB attachment causes haraka to shutdown or crash

Sent an email from one server to an email account on haraka MX server. It caused haraka to shutdown or crash.

Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] TypeError: Cannot create property 'pass' on string 'read ECONNRESET'
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at Plugin.exports.get_class (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:539:23)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at Plugin.exports.format_any (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:447:20)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at RedisClient.plugin.redis.on (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:258:16)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at emitThree (events.js:116:13)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at RedisClient.emit (events.js:197:7)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at return_pub_sub (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:789:18)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at RedisClient.return_reply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:828:9)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at JavascriptRedisParser.returnReply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:192:18)
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at JavascriptRedisParser.execute (/usr/local/lib/node_modules/Haraka/node_modules/redis-parser/lib/parser.js:574:12)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] TypeError: Cannot create property 'pass' on string 'read ECONNRESET'
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at Plugin.exports.get_class (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:539:23)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at Plugin.exports.format_any (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:447:20)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at RedisClient.plugin.redis.on (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:258:16)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at emitThree (events.js:116:13)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at RedisClient.emit (events.js:197:7)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at return_pub_sub (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:789:18)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at RedisClient.return_reply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:828:9)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at JavascriptRedisParser.returnReply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:192:18)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at JavascriptRedisParser.execute (/usr/local/lib/node_modules/Haraka/node_modules/redis-parser/lib/parser.js:574:12)
Jun 8 06:28:36 haraka haraka[45476]: [CRIT] [-] [core] at Socket. (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:274:27)
Jun 8 06:28:36 haraka haraka[45476]: [ERROR] [-] [core] client.socket.remoteAddress undefined!
Jun 8 06:28:36 haraka haraka[45476]: [NOTICE] [-] [core] Shutting down
Jun 8 06:28:36 haraka haraka[45477]: [CRIT] [-] [core] at Socket. (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:274:27)
Jun 8 06:28:36 haraka haraka[45477]: [NOTICE] [-] [core] Shutting down
Jun 8 06:28:39 haraka haraka[45237]: [CRIT] [-] [core] TypeError: Cannot create property 'pass' on string 'read ECONNRESET'
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at Plugin.exports.get_class (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:539:23)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at RedisClient.emit (events.js:197:7)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at return_pub_sub (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:789:18)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at RedisClient.return_reply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:828:9)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at JavascriptRedisParser.returnReply (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:192:18)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at JavascriptRedisParser.execute (/usr/local/lib/node_modules/Haraka/node_modules/redis-parser/lib/parser.js:574:12)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at Socket. (/usr/local/lib/node_modules/Haraka/node_modules/redis/index.js:274:27)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at RedisClient.plugin.redis.on (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:258:16)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at emitThree (events.js:116:13)
Jun 8 06:28:40 haraka haraka[45237]: [CRIT] [-] [core] at Plugin.exports.format_any (/usr/local/lib/node_modules/Haraka/node_modules/haraka-plugin-watch/index.js:447:20)
Jun 8 06:28:40 haraka haraka[45237]: [NOTICE] [-] [core] Shutting down

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,7 +1,8 @@
 language: node_js
 node_js:
-    - "4"
     - "6"
+    - "7"
+    - "8"
 
 before_script:
 

--- a/Changes.md
+++ b/Changes.md
@@ -1,4 +1,9 @@
 
+### 1.0.8 - 2017-06-08
+
+- handle malformed haraka results
+- enable haraka-results publishing #14
+
 ### 1.0.7 - 2017-03-31
 
 - add default config/watch.ini file 

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -2,6 +2,7 @@ environment:
   nodejs_version: "6"
 
 install:
+  - ps: Install-Product node $env:nodejs_version
   - npm install
 
 before_build:

--- a/index.js
+++ b/index.js
@@ -152,6 +152,11 @@ exports.redis_subscribe_all_results = function (next) {
 
       var m = JSON.parse(message);
 
+      if (typeof m.result !== 'object') {
+        plugin.logerror(`garbage was published on ${channel}: ${m.result}`);
+        return;
+      }
+
       switch (m.plugin) {
         case 'local':
           if (m.result.port) {

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "haraka-plugin-watch",
-  "version": "1.0.7",
+  "version": "1.0.8",
   "description": "Watch live SMTP traffic in a web interface",
   "main": "index.js",
   "scripts": {
@@ -27,6 +27,7 @@
   "homepage": "https://github.com/haraka/haraka-plugin-watch#readme",
   "dependencies": {
     "async": "^2.0",
+    "haraka-plugin-redis": "^1.0.5",
     "ws": "^3.0.0"
   },
   "devDependencies": {

--- a/test/watch.js
+++ b/test/watch.js
@@ -5,6 +5,7 @@ var fixtures = require('haraka-test-fixtures');
 describe('watch', function () {
   it('register', function (done) {
     var plugin = new fixtures.plugin('watch');
+    plugin.server = { notes: {} }
     plugin.register();
     assert.ok(plugin.cfg.main);
     // console.log(plugin.cfg);
@@ -13,16 +14,31 @@ describe('watch', function () {
 
   it('loads watch.ini', function (done) {
     var plugin = new fixtures.plugin('watch');
+    plugin.server = { notes: {} }
     plugin.load_watch_ini();
     assert.equal(plugin.cfg.main.sampling, false);
     done();
   });
 
-  it.skip('mocks up a wss socket', function (done) {
+  it('inherits from haraka-plugin-redis', function (done) {
+    var plugin = new fixtures.plugin('watch');
+    plugin.inherits('haraka-plugin-redis');
+    assert.ok(plugin.get_redis_sub_channel);
+    // console.log(plugin);
     done();
   });
 
-  it.skip('inherits from haraka-plugin-redis', function (done) {
+  it('ignores results that are not objects', function (done) {
+    var plugin = new fixtures.plugin('watch');
+    plugin.server = { notes: {} }
+    plugin.register();
+    plugin.load_redis_ini();
+
+    // TODO
+    done();
+  });
+
+  it.skip('mocks up a wss socket', function (done) {
     done();
   });
 
@@ -37,4 +53,5 @@ describe('watch', function () {
   it.skip('sees those results via redis connection subscription', function (done) {
     done();
   });
+
 });

